docs/adr/adr_backend_etl_and_hosting_for_StatViz.md

```diff
@@ -0,0 +1,237 @@
+  `record_id` int(11) DEFAULT NULL,
+  `changes` text,
+  `user_id` int(11) unsigned DEFAULT NULL,
+  `ip` varchar(20) DEFAULT NULL,
+  `changedate` datetime DEFAULT NULL,
+  `from_int` int(11) DEFAULT NULL,
+  `to_int` int(11) DEFAULT NULL,
+  `from_float` float DEFAULT NULL,
+  `to_float` float DEFAULT NULL,
+)
+```
+
+For instance, when a box transitions from one warehouse to an external location, the history table captures it as:
+
+```
+(..., tablename, record_id, changes, ..., changedate, from_int, to_int, ...)
+(..., "stock", "Box1", "location_id", ..., "2023-08-07 12:00:00", 1, 2) // change from WH1 to Free Shop
+(..., "stock", "Box1", "box_state_id", ..., "2023-08-07 12:00:00", 1, 2) // change from InStock to Donated
+```
+
+This data structure efficiently captures transactional data changes, but extracting historical snapshots (like warehouse stock at a particular time) is more complex.
+
+Therefore, we considered introducing shadow tables in the db which at the end would be similar to B.3 with the cache being placed in the db.
+
+## Decisions
+
+1. **[Data Format] Incorporation of User Filters**:We will integrate User Filters as dimensions in the returned data.
+
+   We assume that including these dimensions will not bloat the data to a size where the Front End cannot handle it. No reloads and dynamic filtering for the user are the attractive factors for this decision.
+
+2. **[Data Format] Handling of URL Filters**:URL Filters will be implemented as variables in the GraphQL queries rather than dimensions in the returned data.
+
+   Visualizations are typically embedded via URLs. This design means users won't have dynamic control over these filter values within embedded visualizations. Consequently, embedding URL filters as query variables suffices without burdening the system with additional dimensions that offer no direct user benefit.
+
+3. **[Data Transformation] Nightly Transformations**:
+
+   With our visualizations accessible via a public endpoint and the anticipation of embedding them on public channels, there's potential for a rapid increase in backend requests. This might peak especially during promotional campaigns by our users. To prevent any disruption to the Boxtribute app:
+
+   - We propose hosting the public GraphQL API endpoint on a distinct Google service, separate from the Boxtribute backend.
+   - Data will undergo nightly transformation and be cached independently. Consequently, requests to the public endpoint will bypass resources that the Boxtribute app directly accesses.
+
+4. **[Data Transformation] Caching in DB**:
+   We aim to cache this data in dedicated database tables to avoid the added complexity of filesystem or Redis caching. This is cost-effective in comparison to Redis caching. A lingering question, however, is whether to house this data in a distinct database or an isolated MySQL instance on Google Cloud SQL to guarantee no performance disruptions to the Boxtribute app.
```

Comment:
From a BE perspective it's easier to have all data in one database. Otherwise data fetching becomes complicated: at least two DB connections have to be managed (fetch cache data for facts; fetch data for dimensions, e.g. product names). Sure we can duplicate certain tables for dimensions in the cache DB but that's prone to inconsistencies.

Reply:
good point. haven't thought of that yet.